src/main/scala/inox/ast/ExprOps.scala

```diff
@@ -227,40 +227,81 @@ class ExprOps private(val trees: Trees)
+        Some(ctor(sgn1, i1 + i2, sz1))
+      case Plus(exLit(_, zero, _), e) if zero == BigInt(0) => Some(e)
+      case Plus(e, exLit(_, zero, _)) if zero == BigInt(0) => Some(e)
+      case Plus(e1, UMinus(e2)) => Some(Minus(e1, e2))
+      case Plus(Plus(e, exLit(sgn1, i1, sz1)), exLit(sgn2, i2, sz2)) if sgn1 == sgn2 && sz1 == sz2 =>
+        Some(Plus(e, ctor(sgn1, i1 + i2, sz1)))
+      case Plus(Plus(exLit(sgn1, i1, sz1), e), exLit(sgn2, i2, sz2)) if sgn1 == sgn2 && sz1 == sz2 =>
+        Some(Plus(ctor(sgn1, i1 + i2, sz1), e))
+
+      case Minus(e, exLit(_, zero, _)) if zero == BigInt(0) => Some(e)
+      case Minus(exLit(_, zero, _), e) if zero == BigInt(0) => Some(UMinus(e))
+      case Minus(exLit(sgn1, i1, sz1), exLit(sgn2, i2, sz2)) if sgn1 == sgn2 && sz1 == sz2 =>
+        Some(ctor(sgn1, i1 - i2, sz1))
+      case Minus(e1, UMinus(e2)) => Some(Plus(e1, e2))
+      case Minus(e1, Minus(UMinus(e2), e3)) => Some(Plus(e1, Plus(e2, e3)))
+
+      case UMinus(exLit(true, x, sz)) => Some(ctor(true, -x, sz))
+      case UMinus(UMinus(x)) => Some(x)
+      case UMinus(Plus(UMinus(e1), e2)) => Some(Plus(e1, UMinus(e2)))
+      case UMinus(Minus(e1, e2)) => Some(Minus(e2, e1))
+
+      case Times(exLit(sgn1, i1, sz1), exLit(sgn2, i2, sz2)) if sgn1 == sgn2 && sz1 == sz2 =>
+        Some(ctor(sgn1, i1 * i2, sz1))
+      case Times(exLit(_, one, _), e) if one == BigInt(1) => Some(e)
+      case Times(exLit(_, mone, _), e) if mone == BigInt(-1) => Some(UMinus(e))
+      case Times(e, exLit(_, one, _)) if one == BigInt(1) => Some(e)
+      case Times(exLit(sgn, zero, sz), _) if zero == BigInt(0) => Some(ctor(sgn, 0, sz))
+      case Times(_, exLit(sgn, zero, sz)) if zero == BigInt(0) => Some(ctor(sgn, 0, sz))
+      case Times(exLit(sgn1, i1, sz1), Times(exLit(sgn2, i2, sz2), t)) if sgn1 == sgn2 && sz1 == sz2 =>
+        Some(Times(ctor(sgn1, i1 * i2, sz1), t))
+      case Times(exLit(sgn1, i1, sz1), Times(t, exLit(sgn2, i2, sz2))) if sgn1 == sgn2 && sz1 == sz2 =>
+        Some(Times(ctor(sgn1, i1 * i2, sz1), t))
+      case Times(exLit(true, i, sz), UMinus(e)) => Some(Times(ctor(true, -i, sz), e))
+      case Times(UMinus(e), exLit(true, i, sz)) => Some(Times(e, ctor(true, -i, sz)))
+      case Times(exLit(sgn1, i1, sz1), Division(e, exLit(sgn2, i2, sz2))) if i2 != BigInt(0) && i1 % i2 == BigInt(0) && sgn1 == sgn2 && sz1 == sz2 =>
+        Some(Times(ctor(sgn1, i1/i2, sz1), e))
+
+      case Division(exLit(sgn1, i1, sz1), exLit(sgn2, i2, sz2)) if i2 != BigInt(0) && sgn1 == sgn2 && sz1 == sz2 =>
+        Some(ctor(sgn1, i1 / i2, sz1))
+      case Division(e, exLit(_, one, _)) if one == BigInt(1) => Some(e)
+
+      case _ => None
+    }
+    val bigIntSimplifier = integerArithSimp(ExBigIntLit, (_, l, _) => IntegerLiteral(l))
+    val bvSimplifier = integerArithSimp(ExBVLit, BVLiteral.apply)
+
+    object ExIntegerArithSimp {
+      def unapply(e: Expr): Option[Expr] =
+        bigIntSimplifier(e).orElse(bvSimplifier(e))
```

Comment:
Maybe you could first check whether `e` has an integer or bitvector type to dispatch to the right simplifier. That would also allow you to drop all the sign and size checks since you would know the expression is well-typed.

Reply:
Good point!